scapy/contrib/pcom.py

```diff
@@ -0,0 +1,249 @@
+    0x84: "Read Data Table Reply",
+    0x44: "Write Data Table Request",
+    0xc4: "Write Data Table Reply"
+}
+
+
+class PCOM(Packet):
+    fields_desc = [
+        XShortField("transId", RandShort()),
+        ByteEnumField("mode", 0x65, _protocol_modes),
+        XByteField("reserved", 0x00),
+        LEShortField("len", None)
+    ]
+
+    def post_build(self, pkt, pay):
+        if self.len is None and pay:
+            pkt = pkt[:4] + pack("H", len(pay))
+        return pkt + pay
+
+
+class PCOMRequest(PCOM):
+    name = "PCOM/TCP Request"
+
+    def guess_payload_class(self, payload):
+        if self.mode == 0x65:
+            return PCOMAsciiRequest
+        elif self.mode == 0x66:
+            return PCOMBinaryRequest
+
+
+class PCOMResponse(PCOM):
+    name = "PCOM/TCP Response"
+
+    def guess_payload_class(self, payload):
+        if self.mode == 0x65:
+            return PCOMAsciiResponse
+        elif self.mode == 0x66:
+            return PCOMBinaryResponse
+
+
+class PCOMAscii(Packet):
+    @staticmethod
+    def pcom_ascii_checksum(command):
+        n = 0
+        command = bytes_encode(command)
+        for _, c in enumerate(command):
+            c = c if isinstance(c, int) else ord(c)  # python 2 fallback
+            n += c
+        return list(map(ord, hex(n % 256)[2:].zfill(2).upper()))
+
+
+class PCOMAsciiCommandField(StrLenField):
+    def i2repr(self, pkt, x):
+        s = super(PCOMAsciiCommandField, self).i2repr(pkt, x)
+        code = s[1:4]  # check for 3 chars known codes
+        if code in _ascii_command_codes:
+            return _ascii_command_codes[code] + " " + s
+        code = s[1:3]  # check for 2 chars known codes
+        if code in _ascii_command_codes:
+            return _ascii_command_codes[code] + " " + s
+        return s
+
+
+class PCOMAsciiRequest(PCOMAscii):
+    name = "PCOM/ASCII Request"
+    fields_desc = [
+        StrFixedLenField("stx", "/", 1),
+        StrFixedLenField("unitId", "00", 2),
+        PCOMAsciiCommandField(
+            "command", '', length_from=lambda pkt: pkt.underlayer.len - 6),
+        XShortField("chksum", None),
+        XByteField("etx", 0x0d)
+    ]
+
+    def post_build(self, pkt, pay):
+        if self.chksum is None:
+            chksum = PCOMAscii.pcom_ascii_checksum(pkt[1:-3])
+            pkt = pkt[:-3] + pack("2B", chksum[0], chksum[1]) + pkt[-1:]
+        return pkt + pay
+
+
+class PCOMAsciiResponse(PCOMAscii):
+    name = "PCOM/ASCII Response"
+    fields_desc = [
+        StrFixedLenField("stx", "/A", 2),
+        StrFixedLenField("unitId", "00", 2),
+        PCOMAsciiCommandField(
+            "command", '', length_from=lambda pkt: pkt.underlayer.len - 7),
+        XShortField("chksum", None),
+        XByteField("etx", 0x0d)
+    ]
+
+    def post_build(self, pkt, pay):
+        if self.chksum is None:
+            chksum = PCOMAscii.pcom_ascii_checksum(pkt[2:-3])
+            pkt = pkt[:-3] + pack("2B", chksum[0], chksum[1]) + pkt[-1:]
```

Comment:
If I'm not mistaken, all the calls to `pcom_ascii_checksum` are followed by the same `struct.pack` call. The latter could be moved directly in `pcom_ascii_checksum`